app/Helpers/database/achievement.php

```diff
@@ -27,55 +27,77 @@ function getAchievementsList(
     $bindings = [
         'offset' => $offset,
         'limit' => $limit,
+        'achievementFlag' => $achievementFlag,
     ];
 
-    $innerJoinPlayerAchievements = "";
-    $withAwardedDate = "";
-    if ($params > 0 && $user) {
-        $innerJoinPlayerAchievements = "LEFT JOIN player_achievements AS pa ON pa.achievement_id = ach.ID AND pa.user_id = " . $user->id;
-        $withAwardedDate = ", COALESCE(pa.unlocked_hardcore_at, pa.unlocked_at) AS AwardedDate";
-    }
+    $selectAwardedDate = ", NULL AS AwardedDate";
+    $joinPlayerAchievements = "";
+    $additionalWhereClauses = "";
+    $notExistsSubquery = "";
 
     // We can't run a sort on a user's achievements AwardedDate
     // if we don't have a user. Bail from the sort.
     if (($sortBy == 9 || $sortBy == 19) && !$user) {
         $sortBy = 0;
     }
 
+    if ($params === 1) {
+        // Achievements the user has unlocked.
+        $bindings['userId'] = $user->id;
+
+        $joinPlayerAchievements = "INNER JOIN player_achievements AS pa ON pa.achievement_id = ach.ID AND pa.user_id = :userId";
+        $selectAwardedDate = ", COALESCE(pa.unlocked_hardcore_at, pa.unlocked_at) AS AwardedDate";
+        $additionalWhereClauses .= "AND pa.unlocked_at IS NOT NULL ";
+    } elseif ($params === 2) {
+        // Achievements the user hasn't unlocked.
+        $bindings['userId'] = $user->id;
+
+        $notExistsSubquery = "AND NOT EXISTS (
+            SELECT 1 FROM player_achievements pa
+            WHERE pa.achievement_id = ach.ID AND pa.user_id = :userId
+        ) ";
+    }
+
     // TODO slow query (18)
     $query = "SELECT
-                    ach.ID, ach.Title AS AchievementTitle, ach.Description, ach.Points, ach.TrueRatio, ach.type, ach.DateCreated, ach.DateModified, ach.BadgeName, ach.GameID,
-                    gd.Title AS GameTitle, gd.ImageIcon AS GameIcon, gd.ConsoleID, c.Name AS ConsoleName,
-                    ua.User AS Author
-                    $withAwardedDate
-                FROM Achievements AS ach
-                $innerJoinPlayerAchievements
-                INNER JOIN UserAccounts AS ua ON ua.ID = ach.user_id
-                INNER JOIN GameData AS gd ON gd.ID = ach.GameID
-                INNER JOIN Console AS c ON c.ID = gd.ConsoleID ";
-
-    $bindings['achievementFlag'] = $achievementFlag;
-    $query .= "WHERE ach.Flags = :achievementFlag ";
-
-    // 1 = my unlocked achievements
-    // 2 = achievements i haven't unlocked
-    // 3 = official
-    // 5 = unofficial
-    if ($params == 1) {
-        $query .= "AND pa.unlocked_at IS NOT NULL ";
-    } elseif ($params == 2) {
-        $query .= "AND pa.unlocked_at IS NULL ";
-    }
+                ach.ID,
```

Comment:
Can remove the "slow query comment"?

Reply:
I considered this but intentionally left it in. To be honest, I'd say at 600ms - 1s, it's still slow, just considerably less so.